File: 24SevenClaw-contracts/protocol/rbac.ts
/**
 * 24SevenClaw RBAC (역할 기반 접근 제어) 타입 정의
 * 시스템 역할 + 조직 역할 + 권한 매핑
 */

// === 시스템 역할 ===

export type SystemRole = 'superadmin' | 'admin' | 'member' | 'viewer';

// === 조직 역할 ===

export type OrgRole = 'org_admin' | 'org_member' | 'org_viewer';

// === 권한 ===

export type Permission =
  | 'project:create'
  | 'project:read'
  | 'project:update'
  | 'project:delete'
  | 'preset:manage'
  | 'contract:manage'
  | 'user:manage'
  | 'org:manage'
  | 'report:view'
  | 'rbac:manage';

// === 역할별 권한 매핑 ===

export const ROLE_PERMISSIONS: Record<SystemRole, Permission[]> = {
  superadmin: [
    'project:create', 'project:read', 'project:update', 'project:delete',
    'preset:manage', 'contract:manage', 'user:manage', 'org:manage',
    'report:view', 'rbac:manage',
  ],
  admin: [
    'project:create', 'project:read', 'project:update', 'project:delete',
    'preset:manage', 'contract:manage', 'user:manage', 'org:manage',
    'report:view',
  ],
  member: [
    'project:create', 'project:read', 'project:update', 'project:delete',
    'report:view',
  ],
  viewer: [
    'project:read',
    'report:view',
  ],
};

// === 조직 멤버십 ===

export interface OrganizationMembership {
  id: string;
  user_id: string;
  organization_id: string;
  org_role: OrgRole;
  invited_by: string | null;
  joined_at: string;
  is_active: boolean;
}

// === 역할 감사 로그 ===

export interface RoleAuditEntry {
  id: string;
  actor_id: string;
  target_user_id: string | null;
  action: string;
  old_value: string | null;
  new_value: string;
  resource: string | null;
  created_at: string;
}

File: clickeye-api/app/schemas/rbac.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

SystemRole = Literal["superadmin", "admin", "member", "viewer"]
OrgRole = Literal["org_admin", "org_member", "org_viewer"]
Permission = Literal[
    "project:create",
    "project:read",
    "project:update",
    "project:delete",
    "preset:manage",
    "contract:manage",
    "user:manage",
    "org:manage",
    "report:view",
    "rbac:manage",
]


class PermissionsResponse(BaseModel):
    permissions: list[str]
    system_role: str


class UserAdminResponse(BaseModel):
    id: UUID
    email: str
    display_name: str
    system_role: str
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class RoleUpdateRequest(BaseModel):
    system_role: SystemRole


class UserDeleteResponse(BaseModel):
    """사용자 삭제/비활성화 결과.

    - soft: is_active=False 로 비활성화 (레코드 보존)
    - hard: 레코드 물리 삭제 (superadmin 전용)
    """

    user_id: UUID
    mode: Literal["soft", "hard"]
    is_active: bool
    deleted: bool


class OrgMemberAddRequest(BaseModel):
    user_id: UUID
    org_role: OrgRole = "org_member"


class OrgMemberResponse(BaseModel):
    id: UUID
    user_id: UUID
    organization_id: UUID
    org_role: str
    invited_by: UUID | None
    joined_at: datetime
    is_active: bool

    model_config = {"from_attributes": True}


class AuditLogResponse(BaseModel):
    id: UUID
    actor_id: UUID
    target_user_id: UUID | None
    action: str
    old_value: str | None
    new_value: str
    resource: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogFilter(BaseModel):
    actor_id: UUID | None = None
    target_user_id: UUID | None = None
    action: str | None = None
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)
